lib/libbackscrub.cc

```diff
@@ -99,94 +122,125 @@ static normalization_t get_normalization(modeltype_t type) {
 }
 
 // deeplabv3 classes
+// TODO: read from model metadata file
 static const std::vector<std::string> labels = { "background", "aeroplane", "bicycle", "bird", "boat", "bottle", "bus", "car", "cat", "chair", "cow", "dining table", "dog", "horse", "motorbike", "person", "potted plant", "sheep", "sofa", "train", "tv" };
 // label number of "person" for DeepLab v3+ model
 static const size_t cnum = labels.size();
 static const size_t pers = std::distance(labels.begin(), std::find(labels.begin(),labels.end(),"person"));
 
-int init_tensorflow(calcinfo_t &info) {
-	// Deallocate if not done already
-	if (info.backscrub_ctx)
-		drop_tensorflow(info);
+void *bs_maskgen_new(
+	// Required parameters
+	const char *modelname,
+	size_t threads,
+	size_t width,
+	size_t height,
+	// Optional (nullable) callbacks with caller-provided context
+	// ..debug output
+	void (*ondebug)(void *ctx, const char *fmt, va_list ap),
+	// ..after preparing video frame
+	void (*onprep)(void *ctx),
+	// ..after running inference
+	void (*oninfer)(void *ctx),
+	// ..after generating mask
+	void (*onmask)(void *ctx),
+	// ..the returned context
+	void *caller_ctx
+	) {
 	// Allocate context
-	info.backscrub_ctx = new backscrub_ctx_t;
+	backscrub_ctx_t *pctx = new backscrub_ctx_t;
 	// Take a reference so we can write tidy code with ctx.<x>
-	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+	backscrub_ctx_t &ctx = *pctx;
+	// Save callbacks
+	ctx.ondebug = ondebug;
+	ctx.onprep = onprep;
+	ctx.oninfer = oninfer;
+	ctx.onmask = onmask;
+	ctx.caller_ctx = caller_ctx;
 	// Load model
-	ctx.model = tflite::FlatBufferModel::BuildFromFile(info.modelname);
-	TFLITE_MINIMAL_CHECK(ctx.model != nullptr);
+	ctx.model = tflite::FlatBufferModel::BuildFromFile(modelname);
+	if (nullptr == ctx.model) {
+		_dbg(ctx, "error: unable to load model from file: '%s'.\n", modelname);
+		bs_maskgen_delete(pctx);
+		return nullptr;
+	}
 	// Determine model type and normalization values
-	ctx.modeltype = get_modeltype(info.modelname);
+	ctx.modeltype = get_modeltype(modelname);
 	ctx.norm = get_normalization(ctx.modeltype);
 	if (modeltype_t::Unknown == ctx.modeltype) {
-		_dbg(info, "Unknown model type '%s'.\n", info.modelname);
-		return 0;
+		_dbg(ctx, "error: unknown model type '%s'.\n", modelname);
+		bs_maskgen_delete(pctx);
+		return nullptr;
 	}
 	// Build the interpreter
 	tflite::ops::builtin::BuiltinOpResolver resolver;
 	// custom op for Google Meet network
 	resolver.AddCustom("Convolution2DTransposeBias", mediapipe::tflite_operations::RegisterConvolution2DTransposeBias());
 	tflite::InterpreterBuilder builder(*ctx.model, resolver);
 	builder(&ctx.interpreter);
-	TFLITE_MINIMAL_CHECK(ctx.interpreter != nullptr);
+	if (nullptr == ctx.interpreter) {
```

Comment:
```suggestion
	if (!ctx.interpreter) {
```

Reply:
Agreed, I usually rely on truthy-ness of pointers too, this was a result of translating the macro calls, will fix through the code.